subsys/spm/secure_services.c

```diff
@@ -3,36 +3,74 @@
  *
  * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
  */
+#include <errno.h>
 #include <cortex_m/tz.h>
 #include <misc/reboot.h>
 #include <autoconf.h>
+#include <secure_services.h>
 
-#ifdef CONFIG_SPM_SECURE_SERVICES
-#ifdef CONFIG_SPM_SERVICE_REBOOT
-/** @brief Implementation of Secure System Reset
- *
- * Currently, this simply calls the Zephyr System Reset function,
- * issuing a system reboot immediately.
+/*
+ * Secure Entry functions to allow access to secure services from non-secure
+ * firmware.
  *
  * Note: the function will be located in a Non-Secure
  * Callable region of the Secure Firmware Image.
  */
-static void spm_system_reboot(void)
-{
-	sys_reboot(SYS_REBOOT_WARM);
-}
 
-/** @brief Secure Service for issuing a System Reset
- *
- * Secure Entry function for allowing Non-Secure Firmware
- * to request a System Reset.
+#ifdef CONFIG_SPM_SERVICE_RNG
+#ifdef MBEDTLS_CONFIG_FILE
+#include MBEDTLS_CONFIG_FILE
+#else
+#include "mbedtls/config.h"
+#endif /* MBEDTLS_CONFIG_FILE */
+
+#include <mbedtls/platform.h>
+#include <mbedtls/entropy_poll.h>
+
+/** @brief Structure holding the memory required to generate entropy
+ */
+static mbedtls_rng_workbuf_internal rng_workbuf;
+#endif
+
+
+/** Initialze secure services.
  *
- * Note: the function will be located in a Non-Secure
- * Callable region of the Secure Firmware Image.
+ * @return 0 if successful, negative error otherwise.
  */
-void __TZ_NONSECURE_ENTRY_FUNC spm_request_system_reboot(void)
+int spm_secure_services_init(void)
 {
-	spm_system_reboot();
+	int err = 0;
+
+#ifdef CONFIG_SPM_SERVICE_RNG
+	static mbedtls_platform_context platform_ctx = {
+		.p_rnd_workbuf = &rng_workbuf
+	};
+	err = mbedtls_platform_setup(&platform_ctx);
+#endif
+	return err;
+}
+
+
+#ifdef CONFIG_SPM_SERVICE_REBOOT
+__TZ_NONSECURE_ENTRY_FUNC
+void spm_request_system_reboot(void)
+{
+	sys_reboot(SYS_REBOOT_COLD);
 }
 #endif /* CONFIG_SPM_SERVICE_REBOOT */
-#endif /* CONFIG_SPM_SECURE_SERVICES */
+
+
+#ifdef CONFIG_SPM_SERVICE_RNG
+__TZ_NONSECURE_ENTRY_FUNC
+int spm_request_random_number(u8_t *output, size_t len, size_t *olen)
+{
+	int err;
+
+	if (len != MBEDTLS_ENTROPY_MAX_GATHER) {
+		return -EINVAL;
+	}
+
+	err = mbedtls_hardware_poll(&rng_workbuf, output, len, olen);
```

Comment:
just return mbedtls_hardware_poll, no need for err.

Reply:
I prefer having a possible place for a breakpoint on this line as mbedtls_hardware_poll is implemented inside a library (inside a black box). It helps with debugging and it will have no impact on the compiled code

Reply:
I thought you could break on

return mbedtls_hardware_poll.

e.g.

```
b line 73
c
step
finish

```

Reply:
Left it for now.

Reply:
In command line, yes, but not in all GUI-debuggers. The return of this function is fairly complex since we are going into NSC return, as well